File: Algorithm/maximum_subarray.py
def find_maximum_cross_subarray(arr, left, mid, right):
    l_sum = -2 ** 32
    temp = 0
    max_left, max_right, max_sum = left, right, 0
    for i in range(mid, left - 1, -1):
        temp += arr[i]
        if temp > l_sum:
            l_sum = temp
            max_left = i
    r_sum = -2 ** 32
    temp = 0
    for i in range(mid + 1, right + 1):
        temp += arr[i]
        if temp > r_sum:
            r_sum = temp
            max_right = i
    max_sum = l_sum + r_sum
    return max_left, max_right, max_sum


def find_maximum_subarray(arr, left, right):
    if left == right:
        return left, right, arr[left]
    else:
        mid = (left + right) // 2
        l_left, l_right, l_sum = find_maximum_subarray(arr, left, mid)
        r_left, r_right, r_sum = find_maximum_subarray(arr, mid + 1, right)
        m_left, m_right, m_sum = find_maximum_cross_subarray(arr, left, mid, right)
        max_sum = max([l_sum, m_sum, r_sum])
        if max_sum == l_sum:
            return l_left, l_right, l_sum
        elif max_sum == m_sum:
            return m_left, m_right, m_sum
        else:
            return r_left, r_right, r_sum

File: Algorithm/maximum_subarray.cpp
#include <iostream>
#include <climits>
#include <random>

int find_maximum_crossing_subarray(const int arr[], int left, int mid, int right) {
    int left_sum = INT_MIN;
    int temp = 0;
    for (int i = mid; i >= left; --i) {
        temp += arr[i];
        if (temp > left_sum) {
            left_sum = temp;
        }
    }
    temp = 0;
    int right_sum = INT_MIN;
    for (int i = mid + 1; i <= right; ++i) {
        temp += arr[i];
        if (temp > right_sum) {
            right_sum = temp;
        }
    }
    int max_sum = left_sum + right_sum;
    return max_sum;
}

int find_maximum_subarray(int arr[], int left, int right) {
    if (left == right) {
        return arr[left];
    } else {
        int mid = (left + right) / 2;
        int left_sum = find_maximum_subarray(arr, left, mid);
        int mid_sum = find_maximum_crossing_subarray(arr, left, mid, right);
        int right_sum = find_maximum_subarray(arr, mid + 1, right);
        return std::max(std::max(left_sum, mid_sum), right_sum);
    }
}

int main() {
    int arr[10];
    std::random_device rd;
    std::mt19937 generator(rd());
    std::uniform_int_distribution<int> dist(-10, 10);
    for (int &i : arr) {
        i = dist(generator);
    }
    for (const auto &item : arr) {
        std::cout << " " << item;
    }
    int answer = find_maximum_subarray(arr, 0, sizeof(arr) / sizeof(int) - 1);
    std::cout << '\n' << answer << std::endl;
    return 0;
}